File: sdk/csharp/tests/TestHelpers/MockChainClient.cs
using System.Text.Json;
using Seocheon.Sdk.Infrastructure;
using Seocheon.Sdk.Internal.Tx;

namespace Seocheon.Sdk.Tests.TestHelpers;

/// <summary>
/// In-memory mock chain client for testing.
/// </summary>
public class MockChainClient : IChainClient
{
    public bool IsConnected { get; private set; }
    public long CurrentBlockHeight { get; set; } = 17281; // Epoch 1, Window 0

    private readonly Dictionary<string, JsonElement> _queryResponses = new();
    private readonly Dictionary<string, TxResponse> _txResponses = new();

    public Task ConnectAsync(CancellationToken ct = default)
    {
        IsConnected = true;
        return Task.CompletedTask;
    }

    public Task DisconnectAsync()
    {
        IsConnected = false;
        return Task.CompletedTask;
    }

    public void SetQueryResponse(string path, object response)
    {
        var json = JsonSerializer.Serialize(response);
        _queryResponses[path] = JsonDocument.Parse(json).RootElement.Clone();
    }

    public void SetTxResponse(string txHash, TxResponse response)
    {
        _txResponses[txHash] = response;
    }

    public Task<JsonElement> QueryRest(string path, CancellationToken ct = default)
    {
        // Try exact match first, then prefix match
        if (_queryResponses.TryGetValue(path, out var exact))
            return Task.FromResult(exact);

        foreach (var (key, value) in _queryResponses)
        {
            if (path.StartsWith(key) || path.Contains(key.TrimStart('/')))
                return Task.FromResult(value);
        }

        // Default empty response
        var empty = JsonDocument.Parse("{}").RootElement.Clone();
        return Task.FromResult(empty);
    }

    public Task<BroadcastResponse> BroadcastTx(byte[] txBytes, string mode, CancellationToken ct = default)
    {
        return Task.FromResult(new BroadcastResponse
        {
            TxHash = "ABCD1234567890ABCD1234567890ABCD1234567890ABCD1234567890ABCD1234",
            Code = 0,
            RawLog = ""
        });
    }

    public Task<JsonElement> GetLatestBlock(CancellationToken ct = default)
    {
        var block = new
        {
            block = new
            {
                header = new
                {
                    height = CurrentBlockHeight.ToString(),
                    time = "2026-03-03T00:00:00Z",
                    chain_id = "seocheon-testnet-1"
                },
                data = new
                {
                    txs = Array.Empty<string>()
                }
            }
        };

        var json = JsonSerializer.Serialize(block);
        return Task.FromResult(JsonDocument.Parse(json).RootElement.Clone());
    }

    public Task<TxResponse?> GetTx(string txHash, CancellationToken ct = default)
    {
        if (_txResponses.TryGetValue(txHash, out var response))
            return Task.FromResult<TxResponse?>(response);

        // Default successful TX
        return Task.FromResult<TxResponse?>(new TxResponse
        {
            TxHash = txHash,
            Height = CurrentBlockHeight,
            Code = 0,
            RawLog = "",
            Events = []
        });
    }

    public Task<AccountInfo> GetAccountInfo(string address, CancellationToken ct = default)
    {
        return Task.FromResult(new AccountInfo
        {
            AccountNumber = 42,
            Sequence = 7
        });
    }
}

File: sdk/swift/Tests/SeocheonSDKTests/Helpers/MockChainClient.swift
import Foundation
@testable import SeocheonSDK

final class MockChainClient: ChainClient, @unchecked Sendable {
    var connected = false
    var responses: [String: Data] = [:]
    var broadcastResult = BroadcastResponse(txHash: "AABB00", code: 0, rawLog: "")
    var latestBlock = BlockResponse(height: 34560, time: "2025-01-01T00:00:00Z", chainId: "seocheon-1", numTxs: 5)
    var txResult: TxResponse?
    var accountResult = AccountInfo(accountNumber: 1, sequence: 0)
    var queryError: Error?
    var broadcastError: Error?

    func connect() async throws {
        connected = true
    }

    func disconnect() async {
        connected = false
    }

    func isConnected() -> Bool {
        return connected
    }

    func queryREST(path: String) async throws -> Data {
        if let err = queryError { throw err }
        if let data = responses[path] { return data }
        for (key, data) in responses {
            if path.contains(key) { return data }
        }
        return Data("{}".utf8)
    }

    func broadcastTx(txBytes: Data, mode: String) async throws -> BroadcastResponse {
        if let err = broadcastError { throw err }
        return broadcastResult
    }

    func getLatestBlock() async throws -> BlockResponse {
        return latestBlock
    }

    func getTx(txHash: String) async throws -> TxResponse {
        if let tx = txResult { return tx }
        return TxResponse(txHash: txHash, height: 100, code: 0, gasUsed: 50000, gasWanted: 100000, rawLog: "", events: [])
    }

    func getAccountInfo(address: String) async throws -> AccountInfo {
        return accountResult
    }

    func setResponse(path: String, json: String) {
        responses[path] = Data(json.utf8)
    }
}
